src/vs/editor/contrib/message/browser/messageController.ts

```diff
@@ -76,21 +76,24 @@ export class MessageController implements IEditorContribution {
 		}) : undefined;
 		this._messageWidget.value = new MessageWidget(this._editor, position, typeof message === 'string' ? message : this._message!.element);
 
-		// close on blur, cursor, model change, dispose
-		this._messageListeners.add(this._editor.onDidBlurEditorText(() => {
-			if (!this._focus) {
-				this.closeMessage();
+		// close on blur (debounced to allow to tab into the message), cursor, model change, dispose
+		this._messageListeners.add(Event.debounce(this._editor.onDidBlurEditorText, (last, event) => event, 0)(() => {
+			if (this._mouseOverMessage) {
```

Comment:
I believe you could get away with this property. The editor can tell you if it has text or widget focus and because the message is content widget focusing it will mean the editor signal widget focus ([hasWidgetFocus](https://github.com/microsoft/vscode/blob/fb1d79215e20765f3862c54df12553358947f233/src/vs/editor/browser/editorBrowser.ts#L736))

Reply:
Yeah, I think my intent is to preserve as much of the behaviour today, where the message widget should not remain visible too much, but the only exception is when focus goes into the message widget, specifically when there is a link and you tab into the message to execute it:

![Recording 2023-06-28 at 14 51 37](https://github.com/microsoft/vscode/assets/900690/223363f9-39dd-4920-ae45-5bb88f1209f0)

Since the event order without debouncing will be that `blur` is emitted first, we would otherwise eagerly close the message widget, even though it is in the process of receiving focus.